File: web-visualizer/shared/src/protocol.ts
/**
 * Protocol constants matching the Fabric mod (StreamerWebSocketHandler).
 * Text: HELLO <version>, SET_ORIGIN <x0> <y0> <z0> <scale>.
 * Binary: byte type, then type-specific payload (big-endian).
 */
export const MSG_CHUNK_SECTION_SNAPSHOT = 2;
export const MSG_BLOCK_DELTA = 3;
export const MSG_BLOCK_ENTITY = 4;
export const MSG_ENTITY_SPAWN = 5;

export const SECTION_BLOCK_COUNT = 16 * 16 * 16; // 4096
export const SECTION_BIOME_COUNT = 4 * 4 * 4; // 64
export const DEFAULT_MOD_PORT = 25566;

export interface ChunkSectionCoord {
  cx: number;
  cz: number;
  sy: number;
}

export function chunkSectionKey(coord: ChunkSectionCoord): string {
  return `${coord.cx},${coord.cz},${coord.sy}`;
}

File: MicrocosmXR/Assets/Minecraft/Streaming/Protocol.cs
namespace MicrocosmXR.MinecraftStreaming
{
    /// <summary>
    /// Protocol constants matching the Fabric mod (StreamerWebSocketHandler).
    /// Text: HELLO &lt;version&gt;, SET_ORIGIN &lt;x0&gt; &lt;y0&gt; &lt;z0&gt; &lt;scale&gt;.
    /// Binary: byte type, then type-specific payload (big-endian).
    /// </summary>
    public static class Protocol
    {
        public const byte MsgHello = 0;
        public const byte MsgSetOrigin = 1;
        public const byte MsgChunkSectionSnapshot = 2;
        public const byte MsgBlockDelta = 3;
        public const byte MsgBlockEntity = 4;
        public const byte MsgEntitySpawn = 5;

        public const int SectionBlockCount = 16 * 16 * 16; // 4096
        public const int SectionBiomeCount = 4 * 4 * 4; // 64
        public const int DefaultPort = 25566;
    }

    /// <summary>
    /// Chunk section coordinate (cx, cz, section Y index).
    /// </summary>
    public struct ChunkSectionCoord
    {
        public int cx;
        public int cz;
        public int sy;

        public ChunkSectionCoord(int cx, int cz, int sy)
        {
            this.cx = cx;
            this.cz = cz;
            this.sy = sy;
        }

        public override bool Equals(object obj)
        {
            return obj is ChunkSectionCoord other && cx == other.cx && cz == other.cz && sy == other.sy;
        }

        public override int GetHashCode()
        {
            unchecked
            {
                int h = 17;
                h = h * 31 + cx;
                h = h * 31 + cz;
                h = h * 31 + sy;
                return h;
            }
        }
    }
}
